### PR TITLE
Pbf 540

### PR DESCRIPTION
**Pull request by sarita-ooyala**
Hi Angie,

PBF-540 is branch from PBF-538 .It has all the changes you commented on PBF-538, PBF-546 and PBF-540.

except 

In src/akamai/js/akamaiHD_flash.js:
> +      if (currentUrl.replace(/[\?&]_=[^&]+$/,'') != url)
> +      {
> +        currentUrl = url || "";
> +
> +        // bust the chrome caching bug
> +        if (currentUrl.length > 0) 
> +        {
> +          currentUrl = currentUrl + (/\?/.test(currentUrl) ? "&" : "?") + "_=" + getRandomString();
> +        }
> +        urlChanged = true;
> +        loaded = false;
> +        url = "setVideoUrl("+currentUrl+")";
> +      }
> +      if (_.isEmpty(currentUrl)) 
> +      {
> +        this.controller.notify(this.controller.EVENTS.ERROR, { errorcode: 0 }); 
Actually we might support setting empty string as url for ads as a method of unloading the current stream. It might be good to reconsider this condition

**Implementation for this will pushed under PBF-536 code**
_________
In src/akamai/js/akamaiHD_flash.js:
> @@ -201,8 +205,31 @@ require("../../../html5-common/js/utils/constants.js");
>       * @returns {boolean} True or false indicating success
>       */
>      this.setVideoUrl = function(url, encoding) {
> -      /* Should be set to false if stream URL changes */
> -      loaded = false;
> +      var urlChanged = false;
> +      newController=this.controller;
what is this for or why is this being set here?

 **NewController is created because, this.controller is not accessible inside the onCallback() function. **

Thanks

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -54,15 +59,28 @@ gulp.task('init_module', function(callback) {
...
 
-gulp.task('build_flash', function(callback) {
-  exec("ant -file build_flash.xml", function(err) {
-    if (err) console.log("Error occured in building osmf plugin : " + err);
-    else osmf_fn();
-    callback();
-  });
+gulp.task('build_flash_osmf', function(callback) {
+    exec("ant -buildfile build_flash.xml -Dbuild=build/classes build-CCswc", function(err) {
+      if (err) console.log("Error occured in building the CC : " + err);
```
Comment (angieisrani) [inline gulpfile.js]: looks good - i think i would recommend calling the callback either way otherwise you would prevent other modules from building (which are downstream of this task)

File: src/akamai/js/akamaiHD_flash.js [hunk 2]
```diff
@@ -201,8 +204,31 @@ require("../../../html5-common/js/utils/constants.js");
      * @returns {boolean} True or false indicating success
      */
     this.setVideoUrl = function(url, encoding) {
-      /* Should be set to false if stream URL changes */
-      loaded = false;
+      var urlChanged = false;
+      newController=this.controller;
```
Comment (angieisrani) [inline src/akamai/js/akamaiHD_flash.js]: it makes sense to make a variable that you can pass to callbacks but why is this getting set in setVideoUrl api?  Doesn't seem related in any way

Comment (angieisrani) [inline src/akamai/js/akamaiHD_flash.js]: Also newController is a totally public variable - we need to limit its scope

File: src/akamai/js/akamaiHD_flash.js [hunk 3]
```diff
@@ -449,8 +490,97 @@ require("../../../html5-common/js/utils/constants.js");
...
+          else if (item == "seekRange_end") {
+                seekRange_end = eventData[item];
+          }
+        }
+      }
+
+      switch (eventtitle)
+      {
+       case "JSREADY":
+        for (i = 0; i < actionscriptCommandQueue.length; i++) {
```
Comment (angieisrani) [inline src/akamai/js/akamaiHD_flash.js]: This is cool - I would recommend doing a shift instead to clear the queue as actions are compelted

Comment (mitulparekh) [inline src/akamai/js/akamaiHD_flash.js]: @angieisrani , just for my understanding, are you referring to using something like actionscriptCommandQueue.shift() instead here?

Comment (mitulparekh) [inline src/akamai/js/akamaiHD_flash.js]: @angieisrani, never mind. @sarita-ooyala spoke to me about this. :)

**Comment by sarita-ooyala** [comment]
Hi Angie,

I done all changes according to your comment. Please review these changes .

Thanks

File: src/akamai/js/akamaiHD_flash.js [hunk 4]
```diff
@@ -15,7 +15,9 @@ require("../../../html5-common/js/utils/constants.js");
    * Config variables for paths to flash resources.
    */
 
-  var pluginPath;
+  var scriptsList = document.currentScript;
+  var akamaiPath = scriptsList.src;
+  var pluginPath = akamaiPath.slice(0, 1 + akamaiPath.lastIndexOf("/"));
```
Comment (angieisrani) [inline src/akamai/js/akamaiHD_flash.js]: @stephanieblotner Looks like this is going to mean that we must ask customers not to change the filename of the akamaiHD_flash video plugin.  When this plugin is completed we will need to document this

File: src/akamai/js/akamaiHD_flash.js [hunk 5]
```diff
@@ -201,8 +211,29 @@ require("../../../html5-common/js/utils/constants.js");
      * @returns {boolean} True or false indicating success
      */
     this.setVideoUrl = function(url, encoding) {
-      /* Should be set to false if stream URL changes */
-      loaded = false;
+      newController=this.controller;
```
Comment (angieisrani) [inline src/akamai/js/akamaiHD_flash.js]: It's good that this variable no longer global - but why do we set it in setVideoUrl?

Comment (angieisrani) [inline src/akamai/js/akamaiHD_flash.js]: alternatively why not use something like 
var self = this;
Then you would use self.controller in callbacks.  That'll give you access to all this.* functions

File: src/akamai/js/akamaiHD_flash.js [hunk 6]
```diff
@@ -201,8 +211,29 @@ require("../../../html5-common/js/utils/constants.js");
...
+        {
+          currentUrl = currentUrl + (/\?/.test(currentUrl) ? "&" : "?") + "_=" + getRandomString();
+        }
+        urlChanged = true;
+        loaded = false;
+        url = "setVideoUrl("+currentUrl+")";
+      }
+      if (_.isEmpty(currentUrl)) 
+      {
+        this.controller.notify(this.controller.EVENTS.ERROR, { errorcode: 0 });
```
Comment (angieisrani) [inline src/akamai/js/akamaiHD_flash.js]: For ads we do wish to support empty stream url.  Is it possible to not throw an error here?

File: src/akamai/js/akamaiHD_flash.js [hunk 7]
```diff
@@ -246,8 +277,15 @@ require("../../../html5-common/js/utils/constants.js");
      */
     this.load = function(rewind) {
       if (loaded && !rewind) return;
-      this.callToFlash("load("+rewind+")");
-      loaded = true;
```
Comment (angieisrani) [inline src/akamai/js/akamaiHD_flash.js]: was this intentional?  We also want to call load if we are not rewinding

File: src/akamai/js/akamaiHD_flash.js [hunk 8]
```diff
@@ -449,8 +495,98 @@ require("../../../html5-common/js/utils/constants.js");
...
 
-    // Receives a callback from Flash
+    // Receives a callback from Flash 
     onCallback = _.bind(function(data) {
+      OO.log("[AKAMIHD]:onCallback: ", data);
+      var eventtitle =" ";
+
+      for(var key in data) {
+        if (key == "eventtype") {
+             eventtitle = data[key];
```
Comment (angieisrani) [inline src/akamai/js/akamaiHD_flash.js]: spacing is a tad strange here

**Comment by sarita-ooyala** [comment]
Hi Angie ,

Done all changes according to your comment except 

  if (_.isEmpty(currentUrl)) 
+      {
+        this.controller.notify(this.controller.EVENTS.ERROR, { errorcode: 0 }); 
For ads we do wish to support empty stream url. Is it possible to not throw an error here?

Reason-
Checked and confirmed that the Ad playback is not controlled by OSMF. So the function this.setVideoUrl() won't be called during Ad playback, indicating that no change is required as of now.

**Comment by angieisrani** [comment]
Hi Sarita,
If we start creating ads that are akamai HDS assets then this plugin will be used to render ads.  I think it's a pretty low liklihood but it's certainly possible.

File: src/akamai/js/akamaiHD_flash.js [hunk 9]
```diff
@@ -201,8 +211,29 @@ require("../../../html5-common/js/utils/constants.js");
      * @returns {boolean} True or false indicating success
      */
     this.setVideoUrl = function(url, encoding) {
-      /* Should be set to false if stream URL changes */
-      loaded = false;
+      self=this;
```
Comment (angieisrani) [inline src/akamai/js/akamaiHD_flash.js]: Why are you setting self in setVideo Url?  Why not just set it when self is defined?

**Comment by sarita-ooyala** [comment]
Done changes to support empty stream url.

**Comment by angieisrani** [comment]
LGTM
